### PR TITLE
internal: Migrate assists to the structured snippet API, part 4

### PR DESCRIPTION
**Pull request by DropDemBits**
Continuing from #15260

Migrates the following assists:
- `add_turbo_fish`
- `add_type_ascription`
- `destructure_tuple_binding`
- `destructure_tuple_binding_in_subpattern`

I did this a while ago, but forgot to make a PR for the changes until now. 😅

### CODE REVIEW

File: crates/ide-assists/src/handlers/destructure_tuple_binding.rs [hunk 1]
```diff
@@ -142,98 +152,123 @@ enum RefType {
...
+        if self.in_sub_pattern {
+            ted::insert_all_raw(
+                ted::Position::after(self.ident_pat.syntax()),
+                vec![
+                    make::tokens::single_space().into(),
+                    make::token(T![@]).into(),
+                    make::tokens::single_space().into(),
+                    self.tuple_pat.syntax().clone().into(),
+                ],
+            )
```
Comment (DropDemBits) [inline crates/ide-assists/src/handlers/destructure_tuple_binding.rs:208]: This doesn't feel great, as it makes the syntax tree not match with the parsed version (which may be a problem when automatically fixing the formatting of the syntax tree). I'll see if I can make this be a `ted::replace` instead.

**Comment by Veykril** [comment]
Right, we were actively moving assists to use the mutable syntax node API which is at odds with my opinion of them nowadays https://github.com/rust-lang/rust-analyzer/issues/15710 😓 Wonder if we can figure out a solution for this API should we ever move away from them.
But that aside, moving this is fine for now
@bors r+

**Comment by DropDemBits** [comment]
@Veykril

> Right, we were actively moving assists to use the mutable syntax node API which is at odds with my opinion of them nowadays #15710 😓

Haven't seen that issue until now (slowly getting back into contributing in general after a break), but hard agree on on the problems with the mutable syntax tree api. It is a step up from the text-based apis, but still not ideal 😔

Once I'm done with the structured snippet port + a few structured snippet related odds & ends (according to my personal todo list, maybe mid-December?), I do have an interest in tackling things like the syntax quasi-quote api and at the very least an auto-indenter (though we'll see how I feel by then).

> Wonder if we can figure out a solution for this API should we ever move away from them.

The neat thing about how structured snippets are implemented is that `SourceChangeBuilder` only cares about where nodes & tokens are, and then once the edit is finished, the snippet locations are converted into the corresponding `TextRange` & `TextSize` for the protocol conversion layer to handle, i.e. it does not depend on the mutable syntax tree api 😁

In fact, I think it'd be even better with a different node tracking api, as I've run into the issue of "losing" syntax nodes quite a few times (e.g. for `add_derive`, and in the future of supporting placeholder groups for `extract_variable`).

**Comment by Veykril** [comment]
A related issue regarding node tracking https://github.com/rust-lang/rust-analyzer/issues/9649

**Pull request merged**